File: test/qa/task-compact-boundary.test.mjs
import assert from 'node:assert/strict';
import { copyFile, mkdtemp, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import test from 'node:test';
import { assertBoundaryDOM, boundaryScenarios, loadProducerReceipt, remapProducerFrames } from './task-compact-boundary.mjs';
import { parseArgs, scenarios } from './task-state-ordering.mjs';
import { chat } from './task-state-fixture.mjs';

test('producer option is explicit and the original 14-case matrix remains intact', () => {
  assert.equal(scenarios.length * 2, 14);
  assert.equal(boundaryScenarios.length * 2, 8);
  assert.deepEqual(parseArgs(['--evidence-dir', '/tmp/proof', '--producer-dir', '/tmp/producer']), {
    evidenceDir: '/tmp/proof', producerDir: '/tmp/producer' });
  assert.throws(() => parseArgs(['--evidence-dir', '/tmp/proof', '--other', '/tmp/producer']));
  assert.throws(() => parseArgs(['--evidence-dir', '/tmp/proof', '--producer-dir', '--bad']));
});

for (const mode of ['compact-only', 'mixed']) test(`${mode} keeps emitted payload, compact provenance and clock byte-for-byte while remapping routing`, async () => {
  const directory = await mkdtemp(join(tmpdir(), 'task-producer-helper-'));
  try {
    await copyFile(new URL(`../../frontend/test/fixtures/task-compact-boundary/${mode}.json`, import.meta.url), join(directory, `emitted-${mode}.json`));
    const { receipt, sha256 } = await loadProducerReceipt(directory, mode);
    assert.match(sha256, /^[a-f0-9]{64}$/);
    for (const group of [receipt.frames, receipt.enrichmentFrames]) {
      const mapped = remapProducerFrames(group);
      assert.equal(mapped.length, group.length);
      for (const [i, frame] of mapped.entries()) {
        assert.equal(frame.sessionId, chat);
        assert.equal(group[i].sessionId, 'review-chat', 'original receipt immutable');
        assert.deepEqual({ ...frame, sessionId: 'review-chat' }, group[i]);
      }
    }
    assert.equal(receipt.frames[0].data.tasks[0].padding.length, 65536);
    assert.equal(receipt.frames.at(-1).data.tasks.length, mode === 'mixed' ? 2 : 1);
    const rich = receipt.enrichmentFrames[0].data.tasks.find(row => row.task_id === 'task-r');
    assert.equal(rich.task_summary, 'Full description');
    assert.equal(rich.status, 'completed'); assert.equal(rich.raw_status, 'running');
    assert.equal(rich.updated_at, receipt.expectedTask.updated_at);
  } finally { await rm(directory, { recursive: true, force: true }); }
});

test('boundary DOM assertions reject rollback, descriptor loss, incorrect counts and lost mixed prefix', () => {
  const dom = { agentsSelected: true, sidebarPresent: true, sidebarRunning: 0, count: '0/3', agents: [
    { name: 'running', kind: 'th-activity-chip th-activity-chip--ok' },
    { name: 'pending', kind: 'th-activity-chip th-activity-chip--muted' },
    { name: 'marker', kind: 'th-activity-chip th-activity-chip--muted' },
  ] };
  const expected = { title: 'running', status: 'completed', total: 3 };
  assertBoundaryDOM(dom, expected);
  for (const mutate of [d => { d.agents[0].kind = 'th-activity-chip--running'; },
    d => { d.agents[0].name = 'task-r'; }, d => { d.sidebarRunning = 1; },
    d => { d.count = '0/2'; }, d => { d.agents.pop(); }, d => { d.agents[1].name = 'lost'; },
    d => { d.agentsSelected = false; }]) {
    const bad = structuredClone(dom); mutate(bad); assert.throws(() => assertBoundaryDOM(bad, expected));
  }
});

File: internal/session/task_compact_boundary_test.go
package session

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// These receipts are consumed by activityState.taskCompactBoundary.test.ts.
// Normal Go tests compare every emitted frame with the checked-in fixture;
// regeneration is explicit, and only the nondeterministic receipt clock is omitted.
func TestTaskStateOrderingCompactBoundary(t *testing.T) {
	for _, mode := range []string{"compact-only", "mixed"} {
		t.Run(mode, func(t *testing.T) {
			h := newDAGOrderingHarness(t, "bound")
			r := newRecorder(32)
			detach := h.s.Attach(r)
			defer detach()
			capture := func() []map[string]any {
				injectEvent(t, h.s, map[string]any{"type": "state_changed"})
				before, _ := r.await(t, FrameState)
				frames := make([]map[string]any, 0)
				for _, f := range before {
					if f.Kind != FrameExtensionEvent {
						continue
					}
					data := f.Data.(map[string]any)
					if data["name"] != activitySnapshotOrder[0] {
						continue
					}
					payload, _ := json.Marshal(data["data"])
					var fixtureData map[string]any
					_ = json.Unmarshal(payload, &fixtureData)
					delete(fixtureData, "agent_running_count")
					delete(fixtureData, "agent_total_count")
					delete(fixtureData, "running_count")
					delete(fixtureData, "total_count")
					frames = append(frames, map[string]any{"type": "extensionEvent", "sessionId": "review-chat", "name": data["name"], "data": fixtureData})
				}
				return frames
			}
			row := taskOrderingRow("task-r", "running", dagCurrent)
			row["padding"] = strings.Repeat("x", maxActivitySnapshotBytes)
			h.emit(t, activitySnapshotOrder[0], taskOrderingSnapshot(row))
			if mode == "mixed" {
				data := taskOrderingSnapshot(taskOrderingRow("other", "pending", dagCurrent))
				data["truncated_tasks"] = true
				h.emit(t, activitySnapshotOrder[0], data)
			}
			h.emit(t, activitySnapshotOrder[1], dagOrderingSnapshot(dagOrderingRun("r", "completed", dagNewer)))
			frames := capture()
			summary := h.summary()
			if len(frames) < 2 || summary.TaskDigest == nil {
				t.Fatal("missing producer receipts")
			}
			expected := TaskDigestEntry{TaskID: "task-r", Status: "completed", RawStatus: "running", UpdatedAt: dagCurrent}
			found := false
			for _, task := range summary.TaskDigest.Tasks {
				if task == expected {
					found = true
				}
			}
			if !found || !summary.TaskDigest.Truncated {
				t.Fatalf("compact winner lost: %+v", summary.TaskDigest)
			}
			digestJSON, _ := json.Marshal(summary.TaskDigest)
			if len(digestJSON) > maxActivitySnapshotBytes || len(summary.ActivityPair.Task) > maxActivitySnapshotBytes {
				t.Fatal("cache/digest exceeded 64KiB")
			}
			// Minimal attached corrections must not be stored as complete rich replay:
			// equal-version provider data still has to refill the evicted description.
			rich := taskOrderingRow("task-r", "running", dagCurrent)
			rich["name"], rich["task_summary"] = "Refilled task", "Full description"
			data := taskOrderingSnapshot(rich)
			data["truncated_tasks"] = true
			h.emit(t, activitySnapshotOrder[0], data)
			enrichment := capture()
			enriched := false
			for _, task := range taskOrderingRows(t, h.summary().ActivityPair.Task) {
				if task["task_id"] == "task-r" && task["name"] == rich["name"] && task["task_summary"] == rich["task_summary"] && task["status"] == "completed" && task["raw_status"] == "running" && task["updated_at"] == dagCurrent {
					enriched = true
				}
			}
			if !enriched {
				t.Fatal("equal rich enrichment blocked by compact projection")
			}
			summary.TaskDigest.ReceivedAt = "" // Receipt time is not source authority.
			// This cross-client ordering fixture predates scalar authority and covers
			// retained correction rows only; exact scalar wire coverage is separate.
			fixtureDigest := struct {
				Tasks        []TaskDigestEntry `json:"tasks"`
				Truncated    bool              `json:"truncated"`
				RunningCount int               `json:"running_count"`
				TotalCount   int               `json:"total_count"`
			}{summary.TaskDigest.Tasks, summary.TaskDigest.Truncated, 0, 1}
			receipt := map[string]any{"frames": frames, "taskDigest": fixtureDigest, "expectedTask": expected, "enrichmentFrames": enrichment}
			payload, err := json.MarshalIndent(receipt, "", "  ")
			if err != nil {
				t.Fatal(err)
			}
			payload = append(payload, '\n')
			fixture := filepath.Join("..", "..", "frontend", "test", "fixtures", "task-compact-boundary", mode+".json")
			if os.Getenv("UPDATE_TASK_COMPACT_FIXTURES") == "1" {
				if err := os.MkdirAll(filepath.Dir(fixture), 0755); err != nil {
					t.Fatal(err)
				}
				if err := os.WriteFile(fixture, payload, 0644); err != nil {
					t.Fatal(err)
				}
			}
			shipped, err := os.ReadFile(fixture)
			if err != nil {
				t.Fatal(err)
			}
			if !bytes.Equal(shipped, payload) {
				t.Errorf("producer output differs from frontend-consumed fixture %s", fixture)
			}
			if dir := os.Getenv("TASK_COMPACT_RECEIPT_DIR"); dir != "" {
				path := filepath.Join(dir, "emitted-"+mode+".json")
				if err := os.WriteFile(path, payload, 0600); err != nil {
					t.Fatal(err)
				}
				t.Logf("receipt=%s", path)
			}
			latest, _ := json.Marshal(frames[len(frames)-1]["data"])
			var correction struct {
				Tasks     []TaskDigestEntry `json:"tasks"`
				Truncated bool              `json:"truncated_tasks"`
			}
			if err := json.Unmarshal(latest, &correction); err != nil {
				t.Fatal(err)
			}
			found = false
			for _, task := range correction.Tasks {
				if task == expected {
					found = true
				}
			}
			if !found || !correction.Truncated {
				t.Errorf("attached compact correction missing: %s", latest)
			}
		})
	}
}

// A retained rich prefix must not crowd out any representable compact winner.
func TestTaskStateOrderingCompactWinners(t *testing.T) {
	h := newDAGOrderingHarness(t, "bound")
	r := newRecorder(32)
	detach := h.s.Attach(r)
	defer detach()
	ids := []string{"task-r", "task-s", "task-t"}
	rows := make([]map[string]any, 0, len(ids))
	nodes := make([]any, 0, len(ids))
	for _, id := range ids {
		row := taskOrderingRow(id, "running", dagCurrent)
		row["padding"] = strings.Repeat("x", maxActivitySnapshotBytes)
		rows = append(rows, row)
		nodes = append(nodes, map[string]any{"task_id": id, "state": "completed"})
	}
	h.emit(t, activitySnapshotOrder[0], taskOrderingSnapshot(rows...))
	prefix := taskOrderingRow("other", "pending", dagCurrent)
	prefix["padding"] = strings.Repeat("y", maxActivitySnapshotBytes-220)
	data := taskOrderingSnapshot(prefix)
	data["truncated_tasks"] = true
	h.emit(t, activitySnapshotOrder[0], data)
	run := dagOrderingRun("r", "completed", dagNewer)
	run["nodes"] = nodes
	h.emit(t, activitySnapshotOrder[1], dagOrderingSnapshot(run))
	injectEvent(t, h.s, map[string]any{"type": "state_changed"})
	before, _ := r.await(t, FrameState)
	var latest json.RawMessage
	for _, f := range before {
		if f.Kind != FrameExtensionEvent {
			continue
		}
		data := f.Data.(map[string]any)
		if data["name"] == activitySnapshotOrder[0] {
			latest, _ = json.Marshal(data["data"])
		}
	}
	found := make(map[string]bool)
	for _, row := range taskOrderingRows(t, latest) {
		id := row["task_id"].(string)
		if id == "other" {
			continue
		}
		if row["status"] != "completed" || row["raw_status"] != "running" || row["updated_at"] != dagCurrent {
			t.Fatalf("invalid compact correction: %v", row)
		}
		if found[id] {
			t.Fatalf("duplicate correction: %s", id)
		}
		found[id] = true
	}
	for _, id := range ids {
		if !found[id] {
			t.Errorf("missing compact winner %s beside rich prefix", id)
		}
	}
	summary := h.summary()
	digest, _ := json.Marshal(summary.TaskDigest)
	if summary.TaskDigest == nil || !summary.TaskDigest.Truncated || len(digest) > maxActivitySnapshotBytes || len(summary.ActivityPair.Task) > maxActivitySnapshotBytes {
		t.Fatal("cache/digest bound or partial disclosure lost")
	}
	if len(taskOrderingRows(t, summary.ActivityPair.Task)) != 1 {
		t.Fatal("compact projection polluted rich replay")
	}
}
